pinot-query-planner/src/main/java/org/apache/pinot/calcite/rel/rules/PinotAggregateExchangeNodeInsertRule.java

```diff
@@ -294,42 +299,37 @@ private static List<AggregateCall> buildAggCalls(Aggregate aggRel, AggType aggTy
   //   - argList is replaced with rexList
   private static AggregateCall buildAggCall(RelNode input, AggregateCall orgAggCall, List<RexNode> rexList,
       int numGroups, AggType aggType) {
-    String functionName = orgAggCall.getAggregation().getName();
+    SqlAggFunction orgAggFunction = orgAggCall.getAggregation();
+    String functionName = orgAggFunction.getName();
+    SqlKind kind = orgAggFunction.getKind();
+    SqlFunctionCategory functionCategory = orgAggFunction.getFunctionType();
     if (orgAggCall.isDistinct()) {
-      if (functionName.equals("COUNT")) {
+      if (kind == SqlKind.COUNT) {
         functionName = "DISTINCTCOUNT";
-      } else if (functionName.equals("LISTAGG")) {
+        kind = SqlKind.OTHER_FUNCTION;
+        functionCategory = SqlFunctionCategory.USER_DEFINED_FUNCTION;
+      } else if (kind == SqlKind.LISTAGG) {
         rexList.add(input.getCluster().getRexBuilder().makeLiteral(true));
       }
     }
-    AggregationFunctionType functionType = AggregationFunctionType.getAggregationFunctionType(functionName);
-    SqlAggFunction sqlAggFunction;
-    switch (aggType) {
-      case DIRECT:
-        sqlAggFunction = new PinotSqlAggFunction(functionName, null, functionType.getSqlKind(),
-            ReturnTypes.explicit(orgAggCall.getType()), null, functionType.getOperandTypeChecker(),
-            functionType.getSqlFunctionCategory());
-        break;
-      case LEAF:
-        sqlAggFunction = new PinotSqlAggFunction(functionName, null, functionType.getSqlKind(),
-            functionType.getIntermediateReturnTypeInference(), null, functionType.getOperandTypeChecker(),
-            functionType.getSqlFunctionCategory());
-        break;
-      case INTERMEDIATE:
-        sqlAggFunction = new PinotSqlAggFunction(functionName, null, functionType.getSqlKind(),
-            functionType.getIntermediateReturnTypeInference(), null, OperandTypes.ANY,
-            functionType.getSqlFunctionCategory());
-        break;
-      case FINAL:
-        sqlAggFunction = new PinotSqlAggFunction(functionName, null, functionType.getSqlKind(),
-            ReturnTypes.explicit(orgAggCall.getType()), null, OperandTypes.ANY, functionType.getSqlFunctionCategory());
-        break;
-      default:
-        throw new IllegalStateException("Unsupported AggType: " + aggType);
+    SqlReturnTypeInference returnTypeInference = null;
+    RelDataType returnType = null;
+    // Override the intermediate result type inference if it is provided
+    if (aggType.isOutputIntermediateFormat()) {
+      AggregationFunctionType functionType = AggregationFunctionType.getAggregationFunctionType(functionName);
+      returnTypeInference = functionType.getIntermediateReturnTypeInference();
     }
+    if (returnTypeInference == null) {
+      returnType = orgAggCall.getType();
+      returnTypeInference = ReturnTypes.explicit(returnType);
+    }
```

Comment:
It would be ok to move this code to AggregationFunctionType.getIntermediateReturnTypeInference so it can never return null? Or would it break some other invariant?

Reply:
`AggregationFunctionType.getIntermediateReturnTypeInference` can return null when the return type is the same as final return type. Basically we want to directly use the return type if no intermediate return type inference override is provided. Added some comment